### PR TITLE
FIX (CodeAnalyzer): @W-16121713@: Allow project to live under .dot folders

### LINKED ISSUE #1455: [BUG] canner command not picking up files under ".folder" like ".jenkins"
**Description**:

- When running the command `sf scanner run -t "./**/*.cls"`  for the jenkins job's workspace folder for a PR raised gives error:  
`"The specified target wasn't processed by any engines. Use the --engine parameter to select a different engine or specify a different target. Specified target: ./**/*.cls" and no files are scanned even if the .cls files are present in the job's workspace.`
- The same command when run in the actual sfdx project folder it works fine and the classes scan result is shown in console.

Details:

- Operating System. Example: Windows 11
- sf scanner plugin version: 3.24.0
- Salesforce CLI version: @salesforce/cli/2.38.7

### PR DESCRIPTION
**Pull request by stephen-carter-at-sf**
fixes #1455

### CODE REVIEW

File: test/lib/util/PathMatcher.test.ts [hunk 1]
```diff
@@ -164,6 +164,34 @@ describe('PathMatcher', () => {
 				expect(results).to.not.include(targets[4], 'Path wrongly included');
 			});
 		});
+
+		describe('Hidden dot folders', () => {
+			it('When projectDir contains a .dotFolder then the files underneath it are not matched', async () => {
```
Comment (stephen-carter-at-sf) [inline test/lib/util/PathMatcher.test.ts:169]: This passed before and still passes after.

File: test/lib/util/PathMatcher.test.ts [hunk 2]
```diff
@@ -164,6 +164,34 @@ describe('PathMatcher', () => {
...
+					'/some/projectFolder/subFolder/nonDotFolder/b.js',
+					'/some/projectFolder/subFolder/.dotFolder/c.cls',
+					'/some/projectFolder/subFolder/nonDotFolder/d.cls'
+				];
+
+				const results: string[] = await pm.filterPathsByPatterns(targets);
+				expect(results).to.deep.equal([targets[1],targets[3]]);
+			});
+
+			it('When projectDir contains is under a .dotFolder then the projectDir is not excluded', async () => {
```
Comment (stephen-carter-at-sf) [inline test/lib/util/PathMatcher.test.ts:182]: This failed before but now passes.

File: test/lib/util/PathMatcher.test.ts [hunk 3]
```diff
@@ -263,6 +291,38 @@ describe('PathMatcher', () => {
 			it('DOES NOT MATCH paths matching negative patterns but not positive patterns', async () => {
 				expect(await pm.pathMatchesPatterns(targets[4])).to.equal(false, 'Path wrongly not matched.');
 			});
-		})
-	})
+		});
+
+		describe('Hidden dot folders', () => {
+			it('When projectDir contains a .dotFolder then the files underneath it are not matched', async () => {
```
Comment (stephen-carter-at-sf) [inline test/lib/util/PathMatcher.test.ts:297]: This passed before and still passes after.

File: test/lib/util/PathMatcher.test.ts [hunk 4]
```diff
@@ -263,6 +291,38 @@ describe('PathMatcher', () => {
...
+					'/some/projectFolder/subFolder/nonDotFolder/d.cls'
+				];
+
+				expect(await pm.pathMatchesPatterns(targets[0])).to.equal(false);
+				expect(await pm.pathMatchesPatterns(targets[1])).to.equal(true);
+				expect(await pm.pathMatchesPatterns(targets[2])).to.equal(false);
+				expect(await pm.pathMatchesPatterns(targets[3])).to.equal(true);
+			});
+
+			it('When projectDir contains is under a .dotFolder then the projectDir is not excluded', async () => {
```
Comment (stephen-carter-at-sf) [inline test/lib/util/PathMatcher.test.ts:312]: This failed before but now passes.

File: src/lib/actions/AbstractRunAction.ts [hunk 5]
```diff
@@ -90,11 +90,13 @@ export abstract class AbstractRunAction implements Action {
 		const resultsProcessor: ResultsProcessor = this.resultsProcessorFactory.createResultsProcessor(
 			this.display, outputOptions, jsonReturnValueHolder);
 
+		const primaryProjectDir: string = this.inputProcessor.resolveProjectDirPaths(inputs)[0];
```
Comment (stephen-carter-at-sf) [inline src/lib/actions/AbstractRunAction.ts:93]: This was the cleanest way to get the projectDir all the way into where we want it. Sadly the RuleManager isn't constructed (but is pulled out of thin air) and so sadly, we need to extract out what we want and pass it through all the functions. Thankfully I could just make this an optional parameter so we don't need to update all the tests

File: test/lib/util/PathMatcher.test.ts (hunk 2 above)
Comment (jfeingold35) [inline test/lib/util/PathMatcher.test.ts:182]: "when projectDir contains is under"?

File: test/lib/util/PathMatcher.test.ts (hunk 4 above)
Comment (jfeingold35) [inline test/lib/util/PathMatcher.test.ts:312]: "contains is under"?

File: src/lib/util/PathMatcher.ts [hunk 6]
```diff
@@ -132,6 +134,16 @@ export class PathMatcher {
 	 * @returns {Promise<boolean>}
 	 */
 	public async pathMatchesPatterns(target: string): Promise<boolean> {
-		return this.matcher(normalize(target));
+		return (await this.matcher(normalize(target))) && !this.isDotFileOrIsInDotFolderUnderProjectDir(target);
+	}
+
+	// Note that we do not want to exclude typical files underneath the projectDir just because the projectDir is
```
Comment (jfeingold35) [inline src/lib/util/PathMatcher.ts:140]: Could this be a JSDoc-style comment?

**Review by jfeingold35 — APPROVED** [review]

File: src/lib/util/PathMatcher.ts (hunk 6 above)
Comment (stephen-carter-at-sf) [reply src/lib/util/PathMatcher.ts:140]: Sure... but i'm not sure why... it is a private function.

Comment (stephen-carter-at-sf) [reply src/lib/util/PathMatcher.ts]: Done.

**Review by jfeingold35 — APPROVED** [review]

**Pull request merged**